File: csharp/GardenCatalogue-TestDesign-Kata/GardenCatalogueTest/PlantBuilder.cs
using System.Collections.ObjectModel;
using GardenCatalogue;

namespace GardenCatalogueTest;

public class PlantBuilder
{
    private string _name = "Default Plant";
    private string _latinName = string.Empty;
    private string _articleNumber = string.Empty;
    private PlantType _type = PlantType.Flower;
    private BloomPeriod _bloomPeriod = new BloomPeriod();
    private double _maxHeight = 1.0;
    private SoilCondition _soil = SoilCondition.Any;
    private LightCondition _light = LightCondition.FullSun;
    private Dictionary<string, string> _properties = new();

    public PlantBuilder WithName(string name)
    {
        _name = name;
        return this;
    }

    public PlantBuilder WithLatinName(string latinName)
    {
        _latinName = latinName;
        return this;
    }

    public PlantBuilder WithArticleNumber(string articleNumber)
    {
        _articleNumber = articleNumber;
        return this;
    }

    public PlantBuilder WithType(PlantType type)
    {
        _type = type;
        return this;
    }

    public PlantBuilder WithBloomPeriod(BloomPeriod bloomPeriod)
    {
        _bloomPeriod = bloomPeriod;
        return this;
    }

    public PlantBuilder WithMaxHeight(double maxHeight)
    {
        _maxHeight = maxHeight;
        return this;
    }

    public PlantBuilder WithSoil(SoilCondition soil)
    {
        _soil = soil;
        return this;
    }

    public PlantBuilder WithLight(LightCondition light)
    {
        _light = light;
        return this;
    }

    public PlantBuilder WithProperty(string key, string value)
    {
        _properties[key] = value;
        return this;
    }

    public static PlantBuilder Flower() => new PlantBuilder().WithType(PlantType.Flower);
    public static PlantBuilder Bush() => new PlantBuilder().WithType(PlantType.Bush);
    public static PlantBuilder Tree() => new PlantBuilder().WithType(PlantType.Tree);

    public Plant Build()
    {
        return new Plant(
            _name,
            _latinName,
            _articleNumber,
            _type,
            _bloomPeriod,
            _maxHeight,
            _soil,
            _light,
            new ReadOnlyDictionary<string, string>(_properties));
    }

    public static implicit operator Plant(PlantBuilder builder) => builder.Build();
}

File: java/src/test/java/org/sammancoaching/PlantBuilder.java
package org.sammancoaching;

import java.util.*;

public class PlantBuilder {
    private String name = "Default Plant";
    private String latinName = "";
    private String articleNumber = "";
    private PlantType type = PlantType.Flower;
    private BloomPeriod bloomPeriod = new BloomPeriod(List.of());
    private double maxHeight = 1.0;
    private SoilCondition soil = SoilCondition.Any;
    private LightCondition light = LightCondition.FullSun;
    private Map<String, String> properties = new TreeMap<>();

    public PlantBuilder withName(String name) {
        this.name = name;
        return this;
    }

    public PlantBuilder withLatinName(String latinName) {
        this.latinName = latinName;
        return this;
    }

    public PlantBuilder withArticleNumber(String articleNumber) {
        this.articleNumber = articleNumber;
        return this;
    }

    public PlantBuilder withType(PlantType type) {
        this.type = type;
        return this;
    }

    public PlantBuilder withBloomPeriod(BloomPeriod bloomPeriod) {
        this.bloomPeriod = bloomPeriod;
        return this;
    }

    public PlantBuilder withMaxHeight(double maxHeight) {
        this.maxHeight = maxHeight;
        return this;
    }

    public PlantBuilder withSoil(SoilCondition soil) {
        this.soil = soil;
        return this;
    }

    public PlantBuilder withLight(LightCondition light) {
        this.light = light;
        return this;
    }

    public PlantBuilder withProperty(String key, String value) {
        this.properties.put(key, value);
        return this;
    }

    public static PlantBuilder flower() {
        return new PlantBuilder().withType(PlantType.Flower);
    }

    public static PlantBuilder bush() {
        return new PlantBuilder().withType(PlantType.Bush);
    }

    public static PlantBuilder tree() {
        return new PlantBuilder().withType(PlantType.Tree);
    }

    public Plant build() {
        return new Plant(
                name,
                latinName,
                articleNumber,
                type,
                bloomPeriod,
                maxHeight,
                soil,
                light,
                Collections.unmodifiableMap(new TreeMap<>(properties)));
    }
}
